File: server/src/notifiers/discord.ts
import axios from "axios";
import { AlertPayload } from "../types";
import { INotifier } from "./types";

const COLORS: Record<string, number> = {
  DOWN: 0xff4444,
  DEGRADED: 0xff9900,
  CONTENT_CHANGED: 0x5865f2,
  RECOVERED: 0x2ecc71,
  SSL_EXPIRING: 0xff9900,
  SSL_EXPIRED: 0xff4444,
  SSL_ERROR: 0xff4444,
  SSL_CHANGED: 0x9b59b6,
  CVE_NEW: 0xe74c3c,
};

const discord: INotifier = {
  type: "discord",
  displayName: "Discord",
  configSchema: {
    webhookUrl: {
      label: "Webhook URL",
      type: "text",
      required: true,
      placeholder: "https://discord.com/api/webhooks/...",
    },
  },
  async send(config: Record<string, unknown>, payload: AlertPayload): Promise<void> {
    const webhookUrl = config.webhookUrl as string;
    if (!webhookUrl) throw new Error("Discord webhookUrl is required");

    const fields: { name: string; value: string; inline: boolean }[] = [];

    if (payload.alertType === "CVE_NEW") {
      if (payload.cveDigest && payload.cveDigest.length > 1) {
        // Digest: list top CVEs (Discord allows up to 25 fields)
        const display = payload.cveDigest.slice(0, 23);
        for (const c of display) {
          fields.push({
            name: c.cveId,
            value: `CVSS ${c.cvssScore?.toFixed(1) ?? "N/A"} ${c.cvssSeverity ?? ""}`.trim(),
            inline: true,
          });
        }
        if (payload.cveDigest.length > 23) {
          fields.push({ name: "…and more", value: `+${payload.cveDigest.length - 23} additional CVEs`, inline: false });
        }
      } else {
        if (payload.cveId) {
          fields.push({ name: "CVE ID", value: payload.cveId, inline: true });
        }
        if (payload.cvssScore !== null && payload.cvssScore !== undefined) {
          fields.push({
            name: "CVSS Score",
            value: `${payload.cvssScore.toFixed(1)} ${payload.cvssSeverity ?? ""}`.trim(),
            inline: true,
          });
        }
        fields.push({ name: "NVD Link", value: payload.url, inline: false });
      }
      if (payload.diffViewUrl) {
        fields.push({ name: "View Findings", value: payload.diffViewUrl, inline: false });
      }
    } else {
      fields.push({ name: "URL", value: payload.url, inline: true });
      fields.push({ name: "Status", value: payload.alertType, inline: true });
      if (payload.statusCode !== null) {
        fields.push({ name: "HTTP Status", value: String(payload.statusCode), inline: true });
      }
      if (payload.responseTimeMs !== null) {
        fields.push({ name: "Response Time", value: `${payload.responseTimeMs}ms`, inline: true });
      }
      if (payload.diffViewUrl) {
        fields.push({ name: "View Diff", value: payload.diffViewUrl, inline: false });
      }
    }

    await axios.post(webhookUrl, {
      embeds: [
        {
          title: `[${payload.alertType}] ${payload.serverName}`,
          description: payload.message,
          color: COLORS[payload.alertType] ?? 0x999999,
          fields,
          timestamp: payload.detectedAt,
        },
      ],
    });
  },
};

export { discord };

File: vigil/src/notifiers/discord.rs
use std::collections::HashMap;

use serde_json::{json, Value};

use crate::types::AlertPayload;

pub async fn send(
    config: &HashMap<String, Value>,
    payload: &AlertPayload,
) -> anyhow::Result<()> {
    let webhook_url = config
        .get("webhookUrl")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Discord webhookUrl is required"))?;

    let color: u32 = match payload.alert_type.to_string().as_str() {
        "DOWN" | "SSL_EXPIRED" | "SSL_ERROR" => 0xff4444,
        "DEGRADED" | "SSL_EXPIRING" => 0xff9900,
        "CONTENT_CHANGED" => 0x5865f2,
        "RECOVERED" => 0x2ecc71,
        "SSL_CHANGED" => 0x9b59b6,
        "CVE_NEW" | "CVE_EXPLOIT_ESCALATION" | "CVE_UPDATED" | "CVE_REJECTED" => 0xe74c3c,
        _ => 0x999999,
    };

    let alert_type = payload.alert_type.to_string();
    let mut fields: Vec<Value> = Vec::new();

    let is_cve = matches!(
        alert_type.as_str(),
        "CVE_NEW" | "CVE_EXPLOIT_ESCALATION" | "CVE_UPDATED" | "CVE_REJECTED"
    );

    if is_cve {
        if let Some(digest) = &payload.cve_digest {
            if digest.len() > 1 {
                let take = digest.len().min(23);
                for c in &digest[..take] {
                    let score = c
                        .cvss_score
                        .map(|s| format!("{:.1}", s))
                        .unwrap_or_else(|| "N/A".into());
                    let severity = c.cvss_severity.as_deref().unwrap_or("").to_string();
                    fields.push(json!({
                        "name": c.cve_id,
                        "value": format!("CVSS {} {}", score, severity).trim().to_string(),
                        "inline": true,
                    }));
                }
                if digest.len() > 23 {
                    fields.push(json!({
                        "name": "…and more",
                        "value": format!("+{} additional CVEs", digest.len() - 23),
                        "inline": false,
                    }));
                }
            } else {
                push_cve_fields(&mut fields, payload);
            }
        } else {
            push_cve_fields(&mut fields, payload);
        }
        if let Some(url) = &payload.diff_view_url {
            fields.push(json!({"name": "View Findings", "value": url, "inline": false}));
        }
    } else {
        fields.push(json!({"name": "URL", "value": payload.url, "inline": true}));
        fields.push(json!({"name": "Status", "value": alert_type, "inline": true}));
        if let Some(code) = payload.status_code {
            fields.push(json!({"name": "HTTP Status", "value": code.to_string(), "inline": true}));
        }
        if let Some(ms) = payload.response_time_ms {
            fields.push(json!({"name": "Response Time", "value": format!("{}ms", ms), "inline": true}));
        }
        if let Some(url) = &payload.diff_view_url {
            fields.push(json!({"name": "View Diff", "value": url, "inline": false}));
        }
    }

    let body = json!({
        "embeds": [{
            "title": format!("[{}] {}", payload.alert_type, payload.server_name),
            "description": payload.message,
            "color": color,
            "fields": fields,
            "timestamp": payload.detected_at,
        }]
    });

    reqwest::Client::new()
        .post(webhook_url)
        .json(&body)
        .send()
        .await?
        .error_for_status()?;

    Ok(())
}

fn push_cve_fields(fields: &mut Vec<Value>, payload: &AlertPayload) {
    if let Some(cve_id) = &payload.cve_id {
        fields.push(json!({"name": "CVE ID", "value": cve_id, "inline": true}));
    }
    if let Some(score) = payload.cvss_score {
        let severity = payload.cvss_severity.as_deref().unwrap_or("").to_string();
        fields.push(json!({
            "name": "CVSS Score",
            "value": format!("{:.1} {}", score, severity).trim().to_string(),
            "inline": true,
        }));
    }
    fields.push(json!({"name": "NVD Link", "value": payload.url, "inline": false}));
}
